### ISSUE TITLE
`filter` a lot slower than dplyr

### ISSUE THREAD

**Issue by pdeffebach** [JuliaData/DataFrames.jl#2637]
While debugging some R code I *thought* was slow, I ended up comparing to Julia. 

```
# A tibble: 1,135,821 x 27
   mesh5d  year month total_stays mesh_500m mesh_1000m num_exits_past_…
    <dbl> <dbl> <dbl>       <dbl> <chr>     <chr>                 <dbl>
 1 3.62e9  2014     4           4 362441021 36244102                  0
 2 3.62e9  2014     4           1 362441021 36244102                  0
 3 3.62e9  2019     4           1 362441021 36244102                  0
 4 3.62e9  2014     4           4 362441021 36244102                  0
 5 3.62e9  2015     4           2 362441021 36244102                  0
 6 3.62e9  2016     4           1 362441021 36244102                  0
 7 3.62e9  2017     4           1 362441021 36244102                  0
 8 3.62e9  2018     4           4 362441021 36244102                  0
 9 3.62e9  2014     4          16 362441021 36244102                  0
10 3.62e9  2015     4          36 362441021 36244102                  0
# … with 1,135,811 more rows, and 20 more variables:
#   any_exit_past_year_1000m <lgl>, num_firms_1000m <dbl>,
#   mean_exit_outpref_1000m <dbl>, mean_exit_outown_1000m <dbl>,
#   sum_exit_outpref_1000m <dbl>, sum_exit_outown_1000m <dbl>,
#   num_exits_past_year_500m <dbl>, any_exit_past_year_500m <lgl>,
#   num_firms_500m <dbl>, mean_exit_outpref_500m <dbl>,
#   mean_exit_outown_500m <dbl>, sum_exit_outpref_500m <dbl>,
#   sum_exit_outown_500m <dbl>, num_exits_past_year <dbl>,
#   any_exit_past_year <lgl>, num_firms <dbl>, mean_exit_outpref <dbl>,
#   mean_exit_outown <dbl>, sum_exit_outpref <dbl>, sum_exit_outown <dbl>
0.096 sec elapsed
```

in R we have 

```

r$> tic() 
    filter(bus_stays, total_stays < top_1_pct_stays); 
    toc()    
0.115 sec elapsed
```

In Julia we have 

```
julia> @time filter(:total_stays =>  <(top_1_pct_stays), bus_stays);
  0.740887 seconds (2.27 M allocations: 316.442 MiB)
```

Timings seem unaffected by scoping (which is expected). 

Figured I would report it. It's not clear if this is a copy-on-write thing or not.

**Comment by bkamins** [comment]
What version of PooledArrays.jl do you use (and do you have any categorical or pooled columns)?

**Comment by bkamins** [comment]
In general:

> `2.27 M allocations: 316.442 MiB`

is super bad - not the volume but number of allocations - it shows that some type instability or something other bad thing of this kind happens.

**Comment by pdeffebach** [comment]
Yeah, I think it's a ChainedArrays issue. 

I imported it through Arrow.jl, when I do 

```
df = df[:, :]
``

and use `@btime` I get 

```
julia> global const top_1_pct_stays = quantile(df.total_stays, .99)
julia> @btime filter(:total_stays =>  <(top_1_pct_stays), $df);
  91.169 ms (100 allocations: 247.12 MiB)
```

Which is a little bit faster than R. So I think we are actually good. It was bad benchmarking. 

Not sure what to do about the ChainedArray thing though.

**Comment by bkamins** [comment]
From my experience `ChainedArray` had such problems in the past, probably @quinnj can comment on this.

**Comment by quinnj** [comment]
@pdeffebach, any way you could share the arrow file, even if just privately with me, so I could take a look? We had an issue in the past w/ filter on `ChainedVector`, but it was (supposedly) resolved. I'm always curious in these downstream cases that may suffer from a missing interface method or something and we end up hitting a slow generic fallback.

**Comment by jeremiedb** [comment]
I also observed quite slow filtering operations, which at first sight doesn't seem to be caused by some type instability:

```julia
using DataFrames
df1 = DataFrame(rand(Int(2e6), 200))
@time df2 = filter(:x1 => x -> (x >= 0.2 && x <= 0.8), df1);
 15.675937 seconds (395.40 k allocations: 1.817 GiB, 93.77% gc time)

@time df2 = filter!(:x1 => x -> (x >= 0.2 && x <= 0.8), df1);
1.336566 seconds (373.95 k allocations: 24.030 MiB)
```

With `data.table`: 

```R
library(data.table)
x1 = data.table(matrix(rnorm(2e6 * 200), nrow=2e6))
system.time(x1[V22 >= -0.5 & V22 <= 0.5])
   user  system elapsed 
   0.88    0.78    0.30 
```

The `filter!` shows a decent performance compared to `data.table`. Unfortunately, where the need is to preserve the `df1` object, the `filter` operation brings a significant performance deterioration (over 15X slower than data.table). 

Also, in situations where a filtering was to be applied on the key of a GroupedDataFrame, performance appears good: 
```julia
dfg = groupby(df, :date);
@time dfg2= filter(:date => x -> (x[1] >= start_date), dfg);
0.201490 seconds (165.92 k allocations: 58.881 MiB, 6.65% gc time)
```

However, the resulting `dfg2` cannot apparently be used in downstream task without going through an expensive `DataFrame(dfg2)` since the `transform!` / `select!` & cie operations don't supported subset of GroupedDataFrames: 

```julia
julia> @time select!(dfg , All());
ERROR: ArgumentError: select and transform do not support `GroupedDataFrame`s from which some groups have been dropped (including skipmissing=true)
````

In short, am I missing some beset practices regarding filtering/indexing, or are there some inherent bottlenecks?

**Comment by pdeffebach** [comment]
@jeremiedb check out the [performance tips](https://docs.julialang.org/en/v1/manual/performance-tips/), you are profiling in global scope, which causes a slow-down. 

When I use BenchmarkTools.jl, I get 

```
julia> @btime df2 = filter(:x1 => x -> (x >= 0.2 && x <= 0.8), $df1);
  652.697 ms (617 allocations: 1.80 GiB)
```

When I do the data.table benchmark on my computer I get 

```
r$> library(data.table) 
    x1 = data.table(matrix(rnorm(2e6 * 200), nrow=2e6)) 
    system.time(x1[V22 >= -0.5 & V22 <= 0.5])                              
data.table 1.13.6 using 4 threads (see ?getDTthreads).  Latest news: r-datatable.com
   user  system elapsed 
  0.562   0.498   0.309 
```

So we are within the ballpark of data.table.

**Comment by pdeffebach** [comment]
With regards to your 2nd problem. 

A `GroupedDataFrame` iterates through groups, so `filter` will iterate through groups of a data frame and return a grouped data frame. 

This grouped data frame is "lazy". The underlying data frame is compact, we just only point to a subset of the data now after the filtering instead of all the data. 

So you can't do `transform` because we don't know what to do with the observations that the new grouped data frame no longer points to. 

Honestly, this behavior is pretty unintuitive. The not-yet-released `subset` function has much more intuitive behavior. But it's here to stay since we are finalizing the API.

**Comment by jeremiedb** [comment]
Regarding the filter speed, after fixing the benchmark, I still get quite bad performance: 

```julia
df1 = DataFrame(rand(Int(2e6), 200));
function filter_df_1(df)
    df_f = filter(:x1 => x -> (x >= 0.1 && x <= 0.9), df)
    return df_f
end
@btime filter_df_1($df1);
 15.972 s (617 allocations: 2.40 GiB)
```

I'm wondering it this might not be explained my some gc collection overhead? Although during the benchmark, RAM looked under control: 
![image](https://user-images.githubusercontent.com/18605903/111406546-e2a4a200-86a8-11eb-863b-f520b0f1de56.png)

The above poor speed if also aligned with what I observed running a pre-processing routine which applied the filter within a function call, so it looks like there's something going on outside of the global scope issue.

**Comment by pdeffebach** [comment]
Okay that doesn't make any sense. Your computer must be going to swap for some reason. You should file a new issue with versioninfo and specs so we can track this problem in more detial.

**Comment by jeremiedb** [comment]
Regarding the GroupedDataFrame, would it makes sense that a `filter` returns a GroupedDataFrame whose `parent` is the dataframe filtered on the selected groups? I understand it would involve a reset of the `idx` and such, but it could then be compatible on further `transform/select/combine` operations? 

Maybe it's just a misunderstanding of the docs, but regarding the GroupedDF indexing, is there a "clean" way to filter on the key variables (other than through the integer index)? For example: 
```julia
df1 = DataFrame(a=["a", "b", "a", "c"], b=1:4)
df1g = groupby(df1, :a)

# works but can be combersome for more complexe conditions
df1gf = df1g[[(a = "a",), (a = "c",)]]
```

Is it possible to perform something in the spirit of `dfg[keys(dfg) .∈ Ref(["a", "c"])]`  or `dfg[keys(dfg) .> condition)]` ?

**Comment by pdeffebach** [comment]
> Regarding the GroupedDataFrame, would it makes sense that a filter returns a GroupedDataFrame whose parent is the dataframe filtered on the selected groups?

You should file an issue for this, too. It might be too late to change but it's worth having an issue where we can give a definitive answer and point to in the future. 

> Is it possible to perform something in the spirit of `dfg[keys(dfg) .∈ Ref(["a", "c"])]` or `dfg[keys(dfg) .> condition)]` ?

Probably not. A single `key` is a `NamedTuple`-like object because you can group by multiple columns at once and we want to retain consistency across different kinds of groups. 

I see the appeal, though. For the second example,`dfg[keys(dfg) .> condition)]`,  maybe something like this could live in DataFramesMeta.

**Comment by bkamins** [comment]
Instead of:
```
dfg[keys(dfg) .∈ Ref(["a", "c"])]
```
write something like:
```
dfg[getproperty.(keys(dfg), :a) .∈ Ref(["a", "c"])]
```
and it should work
